src/e2e/liquidity_gap.rs

```diff
@@ -0,0 +1,198 @@
+
+        let dex_x_after = token_x.balance_of(invariant.address());
+        let dex_y_after = token_y.balance_of(invariant.address());
+
+        let delta_dex_x = dex_x_after - dex_x_before;
+        let delta_dex_y = dex_y_before - dex_y_after;
+
+        assert_eq!(user_x, U256::from(0));
+        assert_eq!(user_y, expected_y_amount_out);
+        assert_eq!(delta_dex_x, swap_amount.get());
+        assert_eq!(delta_dex_y, expected_y_amount_out);
+        assert_eq!(
+            pool.fee_growth_global_x,
+            FeeGrowth::new(U128::from(29991002699190242927121u128))
+        );
+        assert_eq!(pool.fee_growth_global_y, FeeGrowth::new(U128::from(0)));
+        assert_eq!(pool.fee_protocol_token_x, TokenAmount::new(U256::from(1)));
+        assert_eq!(pool.fee_protocol_token_y, TokenAmount::new(U256::from(0)));
+    }
+    // No gain swap
+    {
+        let slippage = SqrtPrice::new(U128::from(MIN_SQRT_PRICE));
+        let swap_amount = TokenAmount::new(U256::from(1));
+        let result = invariant.swap(pool_key, true, swap_amount, true, slippage);
+        assert_eq!(result, Err(InvariantError::NoGainSwap));
+    }
+    // Should skip gap and then swap
+    {
+        test_env::set_caller(deployer);
+        let lower_tick_after_swap = -90;
+        let upper_tick_after_swap = -50;
+        let liquidity_delta = Liquidity::from_integer(20008000);
+
+        token_x.mint(&deployer, &liquidity_delta.get());
+        token_y.mint(&deployer, &liquidity_delta.get());
+
+        token_x.approve(invariant.address(), &liquidity_delta.get());
+        token_y.approve(invariant.address(), &liquidity_delta.get());
+
+        let pool_before = invariant
+            .get_pool(pool_key.token_x, pool_key.token_y, fee_tier)
+            .unwrap();
+
+        let slippage_limit_lower = pool_before.sqrt_price;
+        let slippage_limit_upper = pool_before.sqrt_price;
+
+        invariant
+            .create_position(
+                pool_key,
+                lower_tick_after_swap,
+                upper_tick_after_swap,
+                liquidity_delta,
+                slippage_limit_lower,
+                slippage_limit_upper,
+            )
+            .unwrap();
+
+        let caller = test_env::get_account(1);
+        let amount = U256::from(10067);
+        token_x.mint(&caller, &amount);
+
+        test_env::set_caller(caller);
+        token_x.approve(invariant.address(), &amount);
+
+        let slippage = SqrtPrice::new(U128::from(MIN_SQRT_PRICE));
+        let swap_amount = TokenAmount::new(amount);
+        let target_sqrt_price = invariant
+            .quote(pool_key, true, swap_amount, true, slippage)
+            .unwrap()
+            .target_sqrt_price;
+        invariant
+            .swap(pool_key, true, swap_amount, true, target_sqrt_price)
+            .unwrap();
+
+        invariant
+            .get_pool(pool_key.token_x, pool_key.token_y, fee_tier)
+            .unwrap();
```

Comment:
One thing to validate here is to check whether the pool tick is currently in the second position